sql/wsrep_mysqld.cc

```diff
@@ -2881,6 +2880,38 @@ enum wsrep::streaming_context::fragment_unit wsrep_fragment_unit(ulong unit)
   }
 }
 
+bool wsrep_set_encryption_key(const void* key, size_t size, unsigned int version)
+{
+    std::vector<unsigned char> input;
+    wsrep_key_serialize(input, key, size, version);
+    return Wsrep_server_state::instance().set_encryption_key(input);
+}
+
+static const unsigned int key_version_preamble_size = 8;
+
+void wsrep_key_serialize(std::vector<unsigned char>& input, const void* key,
+                         size_t size, unsigned int version)
+{
+  input.resize(key_version_preamble_size + size);
+  memcpy(input.data(), (void *)&version, sizeof(version));
```

Comment:
The key is owned by provider. If provider calls the callback - it means it either wants to encrypt or decrypt some data.

Reply:
This is the point I do not follow: Lets assume gcache is encrypted, based on code every key on file precedes key_version fine. Now, user does shutdown, changes configuration so that binlog  (and gcache) is not encrypted. And restarts the server. So how we determine from gcache that keys are preceded by key_version still ? Similarly, old gcache files, assume binlog encryption is on, how we detect that gcache is not yet encrypted ?

Reply:
If I correctly understand the code we have key_version+encrypted_data on gcache file. Now if I use that file with binlog-encryption=OFF what happens ? Does it try to read that file using "non encrypted" format (FAIL) or somehow notices that file is encrypted (I do not follow how it would notice it) ? You can't trust my.cnf to be correct.